dialogue-core/src/main/java/com/palantir/dialogue/core/BalancedNodeSelectionStrategyChannel.java

```diff
@@ -91,13 +91,25 @@
 
         // TODO(dfox): P2C optimization when we have high number of nodes to save CPU?
         // http://www.eecs.harvard.edu/~michaelm/NEWWORK/postscripts/twosurvey.pdf
-        return preShuffled.stream()
-                .map(MutableChannelWithStats::computeScore)
-                .sorted(Comparator.comparingInt(SortableChannel::getScore))
-                .map(channel -> channel.delegate.maybeExecute(endpoint, request))
-                .filter(Optional::isPresent)
-                .map(Optional::get)
-                .findFirst();
+        List<SortableChannel> sortedByScore = sortByScore(preShuffled);
+
+        for (SortableChannel channel : sortedByScore) {
```

Comment:
If we're going this far, we might as well avoid the iterator as well :-)
```suggestion
        for (int i = 0; i < sortedByScore.size(); i++) {
            SortableChannel channel = sortedByScore.get(i);
```